src/domain/notifications/notifications.repository.v2.ts

```diff
@@ -0,0 +1,163 @@
+   *   registration tokens, or the APNs token has expired for iOS devices).
+   * - If the client app is updated but the new version is not configured to receive
+   *   messages.
+   *
+   * For all these cases, remove this registration token from the app server and stop
+   * using it to send messages.
+   *
+   * @see https://firebase.google.com/docs/cloud-messaging/send-message#rest
+   */
+  static readonly UnregisteredErrorCode = 404;
+  static readonly UnregisteredErrorStatus = 'UNREGISTERED';
+
+  constructor(
+    @Inject(IPushNotificationsApi)
+    private readonly pushNotificationsApi: IPushNotificationsApi,
+    @Inject(INotificationsDatasource)
+    private readonly notificationsDatasource: INotificationsDatasource,
+    @Inject(ISafeRepository)
+    private readonly safeRepository: ISafeRepository,
+    @Inject(IDelegatesV2Repository)
+    private readonly delegatesRepository: IDelegatesV2Repository,
+    @Inject(LoggingService)
+    private readonly loggingService: ILoggingService,
+  ) {}
+
+  async enqueueNotification(args: {
+    token: string;
+    deviceUuid: Uuid;
+    notification: FirebaseNotification;
+  }): Promise<void> {
+    try {
+      await this.pushNotificationsApi.enqueueNotification(
+        args.token,
+        args.notification,
+      );
+    } catch (e) {
+      if (this.isTokenUnregistered(e)) {
+        await this.deleteDevice(args.deviceUuid).catch(() => null);
+      } else {
+        this.loggingService.info(`Failed to enqueue notification: ${e}`);
+        throw new UnprocessableEntityException();
+      }
+    }
+  }
+
+  private isTokenUnregistered(error: unknown): boolean {
+    const isNotFound =
+      get(error, 'code') === NotificationsRepositoryV2.UnregisteredErrorCode;
+    const isUnregistered =
+      get(error, 'status') ===
+      NotificationsRepositoryV2.UnregisteredErrorStatus;
+    return isNotFound && isUnregistered;
+  }
+
+  async upsertSubscriptions(args: UpsertSubscriptionsDto): Promise<{
+    deviceUuid: Uuid;
+  }> {
+    // Only allow owners or delegates to subscribe to notifications
+    // We don't Promise.all getSafe/getDelegates to prevent unnecessary calls
+    for (const safeToSubscribe of args.safes) {
+      const safe = await this.safeRepository
+        .getSafe({
+          chainId: safeToSubscribe.chainId,
+          address: safeToSubscribe.address,
+        })
+        .catch(() => null);
+
+      const isOwner = !!safe?.owners.includes(args.account);
+      if (isOwner) {
+        continue;
+      }
+
+      const delegates = await this.delegatesRepository
+        .getDelegates({
+          chainId: safeToSubscribe.chainId,
+          safeAddress: safeToSubscribe.address,
+          delegate: args.account,
+        })
+        .catch(() => null);
+
+      const isDelegate = !!delegates?.results.some(
+        ({ delegate }) => delegate === args.account,
+      );
+      if (isDelegate) {
+        continue;
+      }
+
+      throw new UnauthorizedException();
```

Comment:
Nit: should we log this error (the account signer not being a owner/delegate of one of the target Safes), wdyt?

Reply:
Added in 663d9016dd7226f0da55c1cef1048f4fa8ef02d3.